jax/_src/lax/control_flow/loops.py

```diff
@@ -1962,12 +1962,20 @@ def fori_loop(lower, upper, body_fun, init_val):
   if use_scan:
     if unroll is None:
       unroll = False
-    if config.disable_jit.value and upper_ == lower_:
-      # non-jit implementation of scan does not support length=0
+    length = upper_ - lower_
+    if length < 0 or (
+        (config.disable_jit.value or unroll is True) and length == 0
+    ):
+      # non-jit and unrolled implementation of scan does not support length <= 0
       return init_val
```

Comment:
Sorry, just noticing this: I think this condition is now too broad. We should continue to bind `scan_p` when `disable_jit = False`, even when the length is zero or negative. The reason for this is that binding the primitive has other effects (like placing the input values on the device and validating the `body` expression) that we should not elide just because the length is zero. This is why we previously did this only in the `disable_jit` case, I think (see https://github.com/google/jax/pull/8160/files#r727120852 for context).

What that suggests to me is that the `unroll` case should probably be handled within the `scan` implementation rather than here – what do you think?

Reply:
Sure I kinda like that:

Inside fori_loop, I have:

```
length = max(upper_ - lower_, 0)
```

Inside scan, I have:

```
  if isinstance(unroll, bool):
    unroll = max(length, 1) if unroll else 1
```